build: minimize cgo

benesch opened the pull request:
One thing led to another and I managed to get this working. With this patch, cgo is no longer used to compile any of our dependencies, save for `engine` and `engineccl` packages. It's a hilariously small change, now that it's all said and done.

A few points:

  * I won't be offended if we decide this isn't the direction we want to take the build system.
  * This compiles on Windows!
  * This compiles on OpenBSD with one tiny patch to make our Go version detection more POSIX-compliant. I'll submit that PR momentarily.
  * The second commit in this PR actually completely bricks the final binary (`Exec format error`) when linking engineccl. Oddly, the binary is just fine with `make buildoss`, even with the second commit. Leaving the commit in the PR in case it's a two-steps-forward-one-step-back kind of deal.
  * [cockroachdb/c-deps](/cockroachdb/c-deps) should probably be replaced with four separate GH repositories that are forked from their upstream sources. Glide actually has no trouble with repositories that contain no Go code. I lumped them all into one repository just to make my life easier.
  * This breaks cross compiling with `xgo`.

benesch committed:
build: minimize cgo's role in the build

This commit excises cgo role in building our dependencies, limiting it
to the link phase only. The c-protobuf, c-jemalloc, c-snappy, and
c-rocksdb packages are removed in favor of Make rules that use these
packages' native build systems. This has a few implications:

  1. We no longer need to maintain platform-specific headers for each
     platform we support; we can rely on our dependency's build systems
     to do feature detection as necessary.

  2. Windows needs no special treatment.

  3. We no longer require the parallel cgo patch for fast compliation
     of C/C++ dependencies.

  4. We, unfortunately, now have a build dependency on CMake. We could
     instead vendor CMake itself, but building CMake from source takes
     about five minutes on my machine and will have serious
     implications for CI times. Binary distributions of CMake are
     readily available, and the specific version of CMake used is not
     particularly critical, as it is with the other C dependencies we
     vendor.

  5. We're no longer compatible with `go get`; you *must* bootstrap
     using Make.

This also adds Windows binaries to the usual deployment scripts.

Closes https://github.com/cockroachlabs/support/issues/33.
Closes https://github.com/cockroachdb/cockroach/issues/14406.

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/14840)
<!-- Reviewable:end -->

petermattis commented:
I'm fine with breaking `go get`, but I believe this change breaks any usage of `go {test,build}` which I'm 👎 on. I frequently `cd <dir>; go test -run SomeTest` or `cd <dir>; go test -run - -bench SomeBench`. I'd really like that to still work, but I don't think that is at odds with this change. Two suggestions:
* Add a `make cdeps` target so a user can easily make sure the C deps have been recompiled instead of making this implicit upon the `build` target.
* Change `build/common.mk` to generate a `.go` file with the appropriate `#cgo` directives instead of setting the `CGO_{LDFLAGS,CPPFLAGS}` environment variables.

tamird commented:
Reviewed 16 of 16 files at r1, 1 of 1 files at r2.
Review status: all files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---

*[build/common.mk, line 86 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14840#-KhXWC5X5gn5zmS7aYSu:-KhXWC5YfJrrZa21Wjki:b-bq7dl7) ([raw file](https://github.com/cockroachdb/cockroach/blob/8b5204eddf9fd8d5970d5df8f478335e337f3b35/build/common.mk#L86)):*
> ```Makefile
> OS_WINDOWS := $(findstring MINGW,$(OS))
> 
> CMAKE_FLAGS = $(if $(OS_WINDOWS),-G "MSYS Makefiles")
> ```

nit: single quotes

---

*[build/common.mk, line 202 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14840#-KhXcEsKwfrkoLTiP0W1:-KhXcEsKwfrkoLTiP0W2:b-rb6kze) ([raw file](https://github.com/cockroachdb/cockroach/blob/8b5204eddf9fd8d5970d5df8f478335e337f3b35/build/common.mk#L202)):*
> ```Makefile
> endif
> 
> PROTOBUF_DIR := $(C_DEPS_DIR)/protobuf/_build
> ```

are these directories git-ignored in the vendor repo? I think I'd prefer if we kept all these builds outside the vendor repository.

---

*[build/common.mk, line 272 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14840#-KhXcRhFXiRaJbLk8HnF:-KhXcRhFXiRaJbLk8HnG:b-ssoc68) ([raw file](https://github.com/cockroachdb/cockroach/blob/8b5204eddf9fd8d5970d5df8f478335e337f3b35/build/common.mk#L272)):*
> ```Makefile
> 	  -DJEMALLOC_INCLUDE_DIR=$(JEMALLOC_INCLUDES) -DJEMALLOC_LIBRARIES=$(LIBJEMALLOC) -DWITH_JEMALLOC=ON
> 
> # C and C++ dependencies must be marked as .ALWAYS_REBUILD.
> ```

explain that their native build systems will do the right thing.

---

*[build/common.mk, line 275 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14840#-KhXcZtlWc3fbwUjpzw1:-KhXcZtlWc3fbwUjpzw2:bdm7hd4) ([raw file](https://github.com/cockroachdb/cockroach/blob/8b5204eddf9fd8d5970d5df8f478335e337f3b35/build/common.mk#L275)):*
> ```Makefile
> 
> $(PROTOC): $(PROTOBUF_DIR)/Makefile .ALWAYS_REBUILD
> 	cd $(PROTOBUF_DIR) && $(MAKE) protoc
> ```

I think we should `-j$(nproc)` all these invocations, otherwise we're gonna have a bad (slow) time.

also, why `cd`? you can `$(MAKE) -C $(DIR) target`.

---

*[build/protobuf.mk, line 73 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14840#-KhXWLNpQ2IZVAriqiCl:-KhXWLNpQ2IZVAriqiCm:b-yxqa9t) ([raw file](https://github.com/cockroachdb/cockroach/blob/8b5204eddf9fd8d5970d5df8f478335e337f3b35/build/protobuf.mk#L73)):*
> ```Makefile
> 	go install $(REPO_ROOT)/vendor/github.com/grpc-ecosystem/grpc-gateway/protoc-gen-grpc-gateway
> 
> $(PROTOC): $(REPO_ROOT)/bin/protoc-gen-gogoroach $(REPO_ROOT)/bin/protoc-gen-grpc-gateway
> ```

eh, do we need to call this `protoc`? it's misleading at best.

---

*[pkg/ccl/storageccl/engineccl/rocksdb.go, line 29 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14840#-KhXWTxWIENfWnrynGd-:-KhXWTxWIENfWnrynGd0:bbei792) ([raw file](https://github.com/cockroachdb/cockroach/blob/8b5204eddf9fd8d5970d5df8f478335e337f3b35/pkg/ccl/storageccl/engineccl/rocksdb.go#L29)):*
> ```Go
> // #cgo linux LDFLAGS: -lrt
> // #cgo windows LDFLAGS: -lrpcrt4
> // #cgo darwin LDFLAGS: -Wl,-undefined -Wl,dynamic_lookup
> ```

we should stick a comment on these now that they're specific to ccl.

ah, you did so in the second commit.

---

*[pkg/ccl/storageccl/engineccl/rocksdb.go, line 22 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14840#-KhXcz1v7xQsSCqPEOf0:-KhXcz1v7xQsSCqPEOf1:bftf22c) ([raw file](https://github.com/cockroachdb/cockroach/blob/dabebe78cb87aecc05734c508a5dbfd491996625/pkg/ccl/storageccl/engineccl/rocksdb.go#L22)):*
> ```Go
> // TODO(tamird): why does rocksdb not link jemalloc,snappy statically?
> 
> // TODO(tamird): why does rocksdb not link jemalloc,snappy statically?
> ```

you accidentally duplicated this TODO

---

*[pkg/ccl/storageccl/engineccl/rocksdb.go, line 36 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14840#-KhXcxSxISBFeyHYDMi9:-KhXcxSxISBFeyHYDMiA:bes3h6t) ([raw file](https://github.com/cockroachdb/cockroach/blob/dabebe78cb87aecc05734c508a5dbfd491996625/pkg/ccl/storageccl/engineccl/rocksdb.go#L36)):*
> ```Go
> // // which aren't linked until the final binary is built. This is the
> // // platform voodoo to make the linker ignore these errors.
> // #cgo windows LDFLAGS: -r
> ```

does `-r` not work for other platforms?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14840)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
> Change build/common.mk to generate a .go file with the appropriate #cgo directives instead of setting the CGO_{LDFLAGS,CPPFLAGS} environment variables.

We used to generate a `pkg-config` file that we'd point to from a `#cgo` directive. We moved away from that to the current scheme because it broke a bunch of tools like `gorename`, `gocode`, etc. I think the problem there was the use of `pkg-config` so a generated `.go` file would be more compatible (even though that solution feels kind of wrong to me for some reason), but in any case we should be testing whatever we're doing here with various tools.

benesch commented:
Pushed an update. Still not perfect, but wanted to get this out today. @petermattis, I haven't fully addressed your concerns, but this revision is again compatible with `go build` and `go test`.

@tamird, C dependencies are now stored in c-deps/FOO.src.tgz. The Makefile expands these tarballs and applies any patches within `c-deps`.

The biggest todo is a patch to invoke `make -jNCPUS` when building our C deps.

---

Review status: 5 of 31 files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---

*[build/common.mk, line 86 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14840#-KhXWC5X5gn5zmS7aYSu:-KhZ94m4_SjcJdXkwCmW:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/8b5204eddf9fd8d5970d5df8f478335e337f3b35/build/common.mk#L86)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

nit: single quotes
</blockquote></details>

Done.

---

*[build/common.mk, line 202 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14840#-KhXcEsKwfrkoLTiP0W1:-KhZ8zigoeF657idvGVq:bhrnwgy) ([raw file](https://github.com/cockroachdb/cockroach/blob/8b5204eddf9fd8d5970d5df8f478335e337f3b35/build/common.mk#L202)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

are these directories git-ignored in the vendor repo? I think I'd prefer if we kept all these builds outside the vendor repository.
</blockquote></details>

Done. Checked in the tarballs to `REPO_ROOT/c-deps`; they build in subdirectories within.

---

*[build/common.mk, line 272 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14840#-KhXcRhFXiRaJbLk8HnF:-KhZ8pnxIASzn8XPYMyh:bg7t1m9) ([raw file](https://github.com/cockroachdb/cockroach/blob/8b5204eddf9fd8d5970d5df8f478335e337f3b35/build/common.mk#L272)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

explain that their native build systems will do the right thing.
</blockquote></details>

Added a couple targets to ease the pain below.

---

*[build/common.mk, line 275 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14840#-KhXcZtlWc3fbwUjpzw1:-KhZ8VFe8t8lpH3_J0Ed:bvy4cto) ([raw file](https://github.com/cockroachdb/cockroach/blob/8b5204eddf9fd8d5970d5df8f478335e337f3b35/build/common.mk#L275)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

I think we should `-j$(nproc)` all these invocations, otherwise we're gonna have a bad (slow) time.

also, why `cd`? you can `$(MAKE) -C $(DIR) target`.
</blockquote></details>

`nproc` is not cross platform. Urgh. The easiest way forward is a Go wrapper that returns `runtime.Ncpus` or whatever it's called. Still TODO.

---

*[build/protobuf.mk, line 73 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14840#-KhXWLNpQ2IZVAriqiCl:-KhZ8Y2pG21RVdZP51v7:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/8b5204eddf9fd8d5970d5df8f478335e337f3b35/build/protobuf.mk#L73)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

eh, do we need to call this `protoc`? it's misleading at best.
</blockquote></details>

Done.

---

*[pkg/ccl/storageccl/engineccl/rocksdb.go, line 29 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14840#-KhXWTxWIENfWnrynGd-:-KhZ8jUmWpPldEbCPz2J:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/8b5204eddf9fd8d5970d5df8f478335e337f3b35/pkg/ccl/storageccl/engineccl/rocksdb.go#L29)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

we should stick a comment on these now that they're specific to ccl.

ah, you did so in the second commit.
</blockquote></details>

Done.

---

*[pkg/ccl/storageccl/engineccl/rocksdb.go, line 22 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14840#-KhXcz1v7xQsSCqPEOf0:-KhZ8ZOJy_UdJJj-FB2I:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/dabebe78cb87aecc05734c508a5dbfd491996625/pkg/ccl/storageccl/engineccl/rocksdb.go#L22)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

you accidentally duplicated this TODO
</blockquote></details>

Done.

---

*[pkg/ccl/storageccl/engineccl/rocksdb.go, line 36 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/14840#-KhXcxSxISBFeyHYDMi9:-KhZ8Zw7GCcpIU2e7nFW:bdnwvx4) ([raw file](https://github.com/cockroachdb/cockroach/blob/dabebe78cb87aecc05734c508a5dbfd491996625/pkg/ccl/storageccl/engineccl/rocksdb.go#L36)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

does `-r` not work for other platforms?
</blockquote></details>

Discussed offline. Doesn't work at all.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14840#-:-KhZ9Kxr489Ysi4tuzHF:b35w0pg)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
Review status: 5 of 31 files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---

*[build/common.mk, line 202 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14840#-KhXcEsKwfrkoLTiP0W1:-Kh_H77fH9l2hGAG1Zau:b8f78h7) ([raw file](https://github.com/cockroachdb/cockroach/blob/8b5204eddf9fd8d5970d5df8f478335e337f3b35/build/common.mk#L202)):*
<details><summary><i>Previously, benesch (Nikhil Benesch) wrote…</i></summary><blockquote>

Done. Checked in the tarballs to `REPO_ROOT/c-deps`; they build in subdirectories within.
</blockquote></details>

I'd prefer to keep the tarballs in a submodule (either `vendor` or `c-deps`. My reasoning is that binary blobs can lead to an ever-growing (and uninteresting except for the ability to reproduce old builds) history. Putting blobs in a submodule gives us more flexibility to abandon that history (or shunt it off to an alternate repo that doesn't have to be downloaded). 

---

*[build/common.mk, line 275 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14840#-KhXcZtlWc3fbwUjpzw1:-Kh_GQyS7UKfNjrtzrni:bwdqy4t) ([raw file](https://github.com/cockroachdb/cockroach/blob/8b5204eddf9fd8d5970d5df8f478335e337f3b35/build/common.mk#L275)):*
<details><summary><i>Previously, benesch (Nikhil Benesch) wrote…</i></summary><blockquote>

`nproc` is not cross platform. Urgh. The easiest way forward is a Go wrapper that returns `runtime.Ncpus` or whatever it's called. Still TODO.
</blockquote></details>

`getconf _NPROCESSORS_ONLN` is POSIX; does it exist everywhere we care about? (it's not on solaris according to [SO](http://stackoverflow.com/a/25565182))

We could also just hardcode a default if decent tools aren't available.

---

*[build/common.mk, line 244 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/14840#-Kh_FAL-alFhGy9WFTu8:-Kh_FAL-alFhGy9WFTu9:b-v0y56m) ([raw file](https://github.com/cockroachdb/cockroach/blob/07007e9a4576267d3b1e0afd59688cf017f797b7/build/common.mk#L244)):*
> ```Makefile
> 
> $(LIBROCKSDB): $(ROCKSDB_DIR)/Makefile $(LIBSNAPPY) $(LIBJEMALLOC)
> 	cd $(ROCKSDB_DIR) && $(MAKE) rocksdb
> ```

What causes these to get rebuilt when we add a new patch or replace a source tarball?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14840)*
<!-- Sent from Reviewable.io -->

benesch commented:
Ok, RFAL!

---

Review status: 5 of 19 files reviewed at latest revision, 8 unresolved discussions, some commit checks failed.

---

*[build/common.mk, line 202 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14840#-KhXcEsKwfrkoLTiP0W1:-Kh_eCQbDiwiIGhaFEsW:bp6stsn) ([raw file](https://github.com/cockroachdb/cockroach/blob/8b5204eddf9fd8d5970d5df8f478335e337f3b35/build/common.mk#L202)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

I'd prefer to keep the tarballs in a submodule (either `vendor` or `c-deps`. My reasoning is that binary blobs can lead to an ever-growing (and uninteresting except for the ability to reproduce old builds) history. Putting blobs in a submodule gives us more flexibility to abandon that history (or shunt it off to an alternate repo that doesn't have to be downloaded). 
</blockquote></details>

Sounds good to me. I created a c-deps/ submodule because having Glide manage these deps in vendor/ felt subtly wrong. (And Glide deletes any files in vendor/ it doesn't recognize, so putting this in vendor means using Glide.)

---

*[build/common.mk, line 275 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/14840#-KhXcZtlWc3fbwUjpzw1:-Kh_fJIJsItyozDk9c3y:b-7fgoi8) ([raw file](https://github.com/cockroachdb/cockroach/blob/8b5204eddf9fd8d5970d5df8f478335e337f3b35/build/common.mk#L275)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

`getconf _NPROCESSORS_ONLN` is POSIX; does it exist everywhere we care about? (it's not on solaris according to [SO](http://stackoverflow.com/a/25565182))

We could also just hardcode a default if decent tools aren't available.
</blockquote></details>

I added a go "one-liner" (pkg/cmd/ncpus) because it's dead simple and guarantees we're covered on any platform that Go supports. Thoughts? `getconf _NPROCESSORS_ONLN` seems like a good alternative and I don't feel strongly.

---

*[build/common.mk, line 244 at r3](https://reviewable.io:443/reviews/cockroachdb/cockroach/14840#-Kh_FAL-alFhGy9WFTu8:-Kh_ew52UZw0KKxxI8bO:b-sgx8s3) ([raw file](https://github.com/cockroachdb/cockroach/blob/07007e9a4576267d3b1e0afd59688cf017f797b7/build/common.mk#L244)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

What causes these to get rebuilt when we add a new patch or replace a source tarball?
</blockquote></details>

Nothing at the revision you reviewed. (I misunderstood Peter's comment, I think.) After talking with Tamir, I've readded `.ALWAYS_REBUILD` so these track dependencies properly. With the newly-added parallelization the overhead is pretty minimal.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14840#-:-Kh_fsWjACtGc6iS9V8u:bkqtryl)*
<!-- Sent from Reviewable.io -->

tamird commented:
Reviewed 4 of 27 files at r3, 27 of 27 files at r4.
Review status: all files reviewed at latest revision, 13 unresolved discussions, some commit checks failed.

---

*[.gitmodules, line 6 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/14840#-Khbklv07RAtlt47ttJm:-Khbklv1VLAy9rTzpFH9:b49i6q6) ([raw file](https://github.com/cockroachdb/cockroach/blob/844b893135d452c02654dc5b6b8515bda9f63162/.gitmodules#L6)):*
> ```
> [submodule "c-deps"]
> 	path = c-deps
> 	url = https://github.com/cockroachdb/c-deps.git
> ```

looking at this repo's import.sh, I wonder:

- should the tarballs be marked as binary via gitattributes?
- https://github.com/cockroachdb/c-deps/blob/master/import.sh#L51:L54 treats all the tarballs as if they're gzip, but jemalloc is bzip2. It happens to work because `tar` (on some platforms) is smart enough to automatically select the correct decompressor, but the code lies.
- what is mangle_rocksdb?

---

*[build/common.mk, line 173 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/14840#-KhbnLHCTlQSrSvBa_55:-KhbnLHCTlQSrSvBa_56:bb7onj7) ([raw file](https://github.com/cockroachdb/cockroach/blob/844b893135d452c02654dc5b6b8515bda9f63162/build/common.mk#L173)):*
> ```Makefile
> 	$(PKG_ROOT)/cmd/metacheck \
> 	$(PKG_ROOT)/cmd/returncheck \
> 	$(PKG_ROOT)/cmd/ncpus \
> ```

sort

---

*[build/common.mk, line 226 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/14840#-KhbnqkvnuG3zcH2Evyz:-KhbnqkvnuG3zcH2Evz-:b-w68hss) ([raw file](https://github.com/cockroachdb/cockroach/blob/844b893135d452c02654dc5b6b8515bda9f63162/build/common.mk#L226)):*
> ```Makefile
> 
> $(PROTOBUF_DIR)/Makefile: $(C_DEPS_DIR)/protobuf.src.tgz | $(PROTOBUF_DIR)
> 	cd $(PROTOBUF_DIR) && cmake $(CMAKE_FLAGS) ./cmake
> ```

just a heads up that in-source builds are generally considered a bad idea in cmake. LLVM outright forbids it.

---

*[build/common.mk, line 250 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/14840#-Khbo-W74BA4f2r36Ms5:-Khbo-W85Uad05_sVuvU:boa4obs) ([raw file](https://github.com/cockroachdb/cockroach/blob/844b893135d452c02654dc5b6b8515bda9f63162/build/common.mk#L250)):*
> ```Makefile
> 
> $(LIBJEMALLOC): $(JEMALLOC_DIR)/Makefile .ALWAYS_REBUILD
> 	$(MAKE) -C $(JEMALLOC_DIR) -j$(NCPUS) build_lib_static LIBPREFIX=lib A=a
> ```

what are these variables?

---

*[build/protobuf.mk, line 68 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/14840#-KhbmRagyauDcu4vKTOR:-KhbmRagyauDcu4vKTOS:b-v0en2y) ([raw file](https://github.com/cockroachdb/cockroach/blob/844b893135d452c02654dc5b6b8515bda9f63162/build/protobuf.mk#L68)):*
> ```Makefile
> 
> $(REPO_ROOT)/bin/protoc-gen-gogoroach: $(BOOTSTRAP_TARGET)
> 	go install $(REPO_ROOT)/pkg/cmd/protoc-gen-gogoroach
> ```

GOINSTALL, right? or whatever we called that variable.

---

*[build/protobuf.mk, line 78 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/14840#-KhbmGTTo-RnDuXj2V2y:-KhbmGTTo-RnDuXj2V2z:b-tinl92) ([raw file](https://github.com/cockroachdb/cockroach/blob/844b893135d452c02654dc5b6b8515bda9f63162/build/protobuf.mk#L78)):*
> ```Makefile
> IMPORT_PREFIX := github.com/$(REPO_NAME)/
> 
> $(GO_SOURCES): protoc-plugins $(GO_PROTOS) $(GOGOPROTO_PROTO)
> ```

this one only needs protoc-gen-gogoroach

---

*[build/protobuf.mk, line 79 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/14840#-KhbmeX4WLOdMNi5Wh6m:-KhbmeX4WLOdMNi5Wh6n:b869dn1) ([raw file](https://github.com/cockroachdb/cockroach/blob/259584fae3763dec5fb3bbda2dd0c82c0be9d498/build/protobuf.mk#L79)):*
> ```Makefile
> IMPORT_PREFIX := github.com/$(REPO_NAME)/
> 
> $(GO_SOURCES): $(PROTOC) $(GO_PROTOS) $(GOGOPROTO_PROTO)
> ```

why did all the targets here lose their dependency on PROTOC?

---

*[build/protobuf.mk, line 90 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/14840#-KhbmJ8SH-2iAMCCoy9s:-KhbmJ8SH-2iAMCCoy9t:b-wzejj0) ([raw file](https://github.com/cockroachdb/cockroach/blob/844b893135d452c02654dc5b6b8515bda9f63162/build/protobuf.mk#L90)):*
> ```Makefile
> 	gofmt -s -w $(GO_SOURCES)
> 
> $(GW_SOURCES): protoc-plugins $(GW_SERVER_PROTOS) $(GW_TS_PROTOS) $(GO_PROTOS) $(GOGOPROTO_PROTO)
> ```

this one only needs protoc-gen-grpc-gateway

---

*[build/protobuf.mk, line 105 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/14840#-KhbmMf9RB50n_X_Lg_7:-KhbmMf9RB50n_X_Lg_8:b-e2y8h5) ([raw file](https://github.com/cockroachdb/cockroach/blob/844b893135d452c02654dc5b6b8515bda9f63162/build/protobuf.mk#L105)):*
> ```Makefile
> 	pbts $(UI_JS) >> $@
> 
> $(CPP_HEADERS) $(CPP_SOURCES): protoc-plugins $(CPP_PROTOS)
> ```

this one doesn't need any plugins

---

*[pkg/ccl/storageccl/engineccl/rocksdb.go, line 35 at r4](https://reviewable.io:443/reviews/cockroachdb/cockroach/14840#-Khbm2pTJY7ujg-oyadN:-Khbm2pTJY7ujg-oyadO:bxuabw8) ([raw file](https://github.com/cockroachdb/cockroach/blob/844b893135d452c02654dc5b6b8515bda9f63162/pkg/ccl/storageccl/engineccl/rocksdb.go#L35)):*
> ```Go
> // // which aren't linked until the final binary is built. This is the
> // // platform voodoo to make the linker ignore these errors.
> // // TODO: Make this work on Windows.
> ```

https://github.com/cockroachdb/cockroach/issues/14673

if you leave the TODO, put a name on it (mine is fine).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14840)*
<!-- Sent from Reviewable.io -->

tamird commented:
@petermattis @benesch many changes made here. Can y'all have a look?

petermattis commented:
@tamird I'm not going to have time to look at this tonight and probably not tomorrow. Hopefully you and @benesch can get it done.

benesch commented:
The upshot is that `go build` and `go test` now work as before, provided
you run `make` once to bootstrap the C dependencies.

N

On Thu, Apr 13, 2017 at 7:57 PM Peter Mattis <notifications@github.com>
wrote:

> @tamird <https://github.com/tamird> I'm not going to have time to look at
> this tonight and probably not tomorrow. Hopefully you and @benesch
> <https://github.com/benesch> can get it done.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/14840#issuecomment-294050975>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AA15IE4GaXCh7S0P-aEn4v8cPbqu7wv2ks5rvraFgaJpZM4M7L6s>
> .
>

tamird commented:
Alright, this is green. Turns out compiling rocksdb with `-O2` is pretty important.

benesch commented:
Whew! @tamird you planning to get cross-compiling working in this PR too?

tamird commented:
Yeah, probs should. Tomorrow.

On Thu, Apr 13, 2017 at 10:29 PM, Nikhil Benesch <notifications@github.com>
wrote:

> Whew! @tamird <https://github.com/tamird> you planning to get
> cross-compiling working in this PR too?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/14840#issuecomment-294070659>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABdsPHZyBnoGVSOpx8o6M6_eXeu5gBTtks5rvtoLgaJpZM4M7L6s>
> .
>

tamird commented:
This is more or less complete (make check fails because of mishandling of build tags, but @benesch is going to fix that shortly).

In addition to using the native build systems, this also includes automation to build and publish windows builds. Hurray!

petermattis commented:
Review status: 6 of 46 files reviewed at latest revision, 14 unresolved discussions, some commit checks failed.

---

*[build/common.mk, line 228 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/14840#-KhyRYe9wwu4P3y4qYFJ:-KhyRYe9wwu4P3y4qYFK:bbtc9wj) ([raw file](https://github.com/cockroachdb/cockroach/blob/b22025fad909f4eb41880b3528477cec381a6ec2/build/common.mk#L228)):*
> ```Makefile
> 
> CGO_FLAGS_FILES := $(addprefix $(PKG_ROOT)/,$(addsuffix /zcgo_flags_$(TARGET_TRIPLE).go,cli server/status storage/engine ccl/storageccl/engineccl))
> $(CGO_FLAGS_FILES):
> ```

Why does `CGO_FLAGS_FILES` need to have `TARGET_TRIPLE` as the suffix? Is that so you can build multiple versions out of the same source directory? 

I think it would suffice to suffix `<GOOS>_<GOARCH>`. I'm imagining you'd regenerate `CGO_FLAGS_FILES` on every invocation (if changed) so that if something about `TARGET_TRIPLE` changed for a particular `GOOS/GOARCH` pair we'd have to recompile, but otherwise we'd be fine. I'm likely missing some requirement here.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14840)*
<!-- Sent from Reviewable.io -->

tamird commented:
Review status: 6 of 46 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[build/common.mk, line 228 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/14840#-KhyRYe9wwu4P3y4qYFJ:-KhyZQ5Q1aStNUloV2SJ:b-z76pg4) ([raw file](https://github.com/cockroachdb/cockroach/blob/b22025fad909f4eb41880b3528477cec381a6ec2/build/common.mk#L228)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

Why does `CGO_FLAGS_FILES` need to have `TARGET_TRIPLE` as the suffix? Is that so you can build multiple versions out of the same source directory? 

I think it would suffice to suffix `<GOOS>_<GOARCH>`. I'm imagining you'd regenerate `CGO_FLAGS_FILES` on every invocation (if changed) so that if something about `TARGET_TRIPLE` changed for a particular `GOOS/GOARCH` pair we'd have to recompile, but otherwise we'd be fine. I'm likely missing some requirement here.
</blockquote></details>

All the dependencies are built out-of-source; that is, we only have one source tree but many build directories (one per triple).

The reason we use `TARGET_TRIPLE` rather than `<GOOS>_<GOARCH>` is that the latter would break in the case where some builds a gnu release followed by a musl release, because the two have the same `<GOOS>_<GOARCH>` but different `TARGET_TRIPLE`s.

You're suggesting we'd regenerate `zcgo_ldflags_<GOOS>_<GOARCH>` if `TARGET_TRIPLE` changed, but how would we detect that? We'd need to add another marker file?

The requirement is that we be able to run `make build TYPE=release && make build TYPE=musl-release` and have correct results; apart from that, I'm willing to entertain any suggestions you have (the current one isn't concrete enough yet).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14840)*
<!-- Sent from Reviewable.io -->

petermattis commented:
Review status: 6 of 46 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[build/common.mk, line 228 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/14840#-KhyRYe9wwu4P3y4qYFJ:-Khyq4mY6FvFoyrJQ-aO:b-f7fnhd) ([raw file](https://github.com/cockroachdb/cockroach/blob/b22025fad909f4eb41880b3528477cec381a6ec2/build/common.mk#L228)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

All the dependencies are built out-of-source; that is, we only have one source tree but many build directories (one per triple).

The reason we use `TARGET_TRIPLE` rather than `<GOOS>_<GOARCH>` is that the latter would break in the case where some builds a gnu release followed by a musl release, because the two have the same `<GOOS>_<GOARCH>` but different `TARGET_TRIPLE`s.

You're suggesting we'd regenerate `zcgo_ldflags_<GOOS>_<GOARCH>` if `TARGET_TRIPLE` changed, but how would we detect that? We'd need to add another marker file?

The requirement is that we be able to run `make build TYPE=release && make build TYPE=musl-release` and have correct results; apart from that, I'm willing to entertain any suggestions you have (the current one isn't concrete enough yet).
</blockquote></details>

I'm thinking of something like this example:

```
all: cgo_flags.go
	@touch $@
	@echo rebuilding

.PHONY: cgo_flags
cgo_flags:
	@echo $(GOOS) > $@.tmp
	@if ! cmp $@.tmp $@.go >/dev/null 2>&1 ; then \
	  mv $@.tmp $@.go; \
	else \
	  rm -f $@.tmp; \
	fi

-include cgo_flags
```

The `-include cgo_flags` rule causes that target to be rebuilt on every make invocation, but we only update `cgo_flags.go` if `$(GOOS)` has changed. When we get to the processing of normal rules we only rebuild the package containing `cgo_flags.go` if `$(GOOS)` has changed. I think this could be made to work if `$(GOOS)` is replaced by `$(TARGET_TRIPLE)`. The result is a rebuild when you switch from target to target, but that seems acceptable. We can get slightly fancier, but let's see if this basic idea works first.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14840)*
<!-- Sent from Reviewable.io -->

tamird commented:
Review status: 6 of 46 files reviewed at latest revision, 2 unresolved discussions, some commit checks failed.

---

*[build/common.mk, line 228 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/14840#-KhyRYe9wwu4P3y4qYFJ:-Ki0VU9vClX3y_KydK3z:b-wnyvs2) ([raw file](https://github.com/cockroachdb/cockroach/blob/b22025fad909f4eb41880b3528477cec381a6ec2/build/common.mk#L228)):*
<details><summary><i>Previously, petermattis (Peter Mattis) wrote…</i></summary><blockquote>

I'm thinking of something like this example:

```
all: cgo_flags.go
	@touch $@
	@echo rebuilding

.PHONY: cgo_flags
cgo_flags:
	@echo $(GOOS) > $@.tmp
	@if ! cmp $@.tmp $@.go >/dev/null 2>&1 ; then \
	  mv $@.tmp $@.go; \
	else \
	  rm -f $@.tmp; \
	fi

-include cgo_flags
```

The `-include cgo_flags` rule causes that target to be rebuilt on every make invocation, but we only update `cgo_flags.go` if `$(GOOS)` has changed. When we get to the processing of normal rules we only rebuild the package containing `cgo_flags.go` if `$(GOOS)` has changed. I think this could be made to work if `$(GOOS)` is replaced by `$(TARGET_TRIPLE)`. The result is a rebuild when you switch from target to target, but that seems acceptable. We can get slightly fancier, but let's see if this basic idea works first.
</blockquote></details>

@benesch is doing something with this.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14840)*
<!-- Sent from Reviewable.io -->

tamird commented:
Review status: 6 of 45 files reviewed at latest revision, 2 unresolved discussions, some commit checks pending.

---

*[build/common.mk, line 228 at r5](https://reviewable.io:443/reviews/cockroachdb/cockroach/14840#-KhyRYe9wwu4P3y4qYFJ:-Ki0oczKOWVZL-QcgQIy:b-jqva5f) ([raw file](https://github.com/cockroachdb/cockroach/blob/b22025fad909f4eb41880b3528477cec381a6ec2/build/common.mk#L228)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

@benesch is doing something with this.
</blockquote></details>

@petermattis PTAL, this should address your needs now.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14840)*
<!-- Sent from Reviewable.io -->

tamird committed:
cmd/github-pull-request-make: os-agnosticize tests

bdarnell commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 5 of 16 files at r1, 1 of 27 files at r4, 35 of 43 files at r5, 4 of 5 files at r6.
Review status: all files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[Makefile, line 45 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/14840#-Ki15vq-1QQl82cSmc9U:-Ki15vq-1QQl82cSmc9V:b-7qowfu) ([raw file](https://github.com/cockroachdb/cockroach/blob/10a642d72aacb5ac2a99f256977783d3bb510c5c/Makefile#L45)):*
> ```Makefile
> # Possible values:
> # <empty>: use the default toolchain
> # release:         target Linux 2.6.32, dynamically link GLIBC 2.12.2
> ```

Can we rename `release` to `release-linux`?

---

*[build/common.mk, line 299 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/14840#-Ki17dnXa-uUeVRu38Ln:-Ki17dnY3IWiTBsMTEsA:b-cer0c2) ([raw file](https://github.com/cockroachdb/cockroach/blob/10a642d72aacb5ac2a99f256977783d3bb510c5c/build/common.mk#L299)):*
> ```Makefile
> 
> $(CGO_FLAGS_FILES):
> 	@echo 'GEN $@'
> ```

Did you mean to remove this line before committing?

---

*[build/push-tagged-aws.sh, line 53 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/14840#-Ki19jYItOtZteE84KtF:-Ki19jYJ-BUFU2KTPLPK:bamzmi7) ([raw file](https://github.com/cockroachdb/cockroach/blob/10a642d72aacb5ac2a99f256977783d3bb510c5c/build/push-tagged-aws.sh#L53)):*
> ```Shell
> 
> # Tar up and push Windows binaries. One tagged, one called "latest". The local
> # binary name is generated by build-osx.sh.
> ```

`build-osx.sh`? I think both references to that script need to point to the Makefile (or just remove these parts of the comments)

---

*[build/teamcity-release-upload.sh, line 43 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/14840#-Ki1AAQikY36waljnuFI:-Ki1AAQikY36waljnuFJ:b-ol5j88) ([raw file](https://github.com/cockroachdb/cockroach/blob/10a642d72aacb5ac2a99f256977783d3bb510c5c/build/teamcity-release-upload.sh#L43)):*
> ```Shell
> aws configure set region us-east-1
> build/build-osx.sh
> build/build-windows.sh
> ```

Don't change it in this PR, but this part of the build process is taking long enough that we should look at parallelizing it soon.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14840#-:-Ki1B5-POC2VyjRtGw-0:bnfp4nl)*
<!-- Sent from Reviewable.io -->

tamird commented:
Review status: 0 of 46 files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[Makefile, line 45 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/14840#-Ki15vq-1QQl82cSmc9U:-Ki1BMyc5CF5rKXad-Qu:b2xvh50) ([raw file](https://github.com/cockroachdb/cockroach/blob/10a642d72aacb5ac2a99f256977783d3bb510c5c/Makefile#L45)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Can we rename `release` to `release-linux`?
</blockquote></details>

should we rename both it and release-musl to `release-linux-gnu` and `release-linux-musl`?

---

*[build/common.mk, line 299 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/14840#-Ki17dnXa-uUeVRu38Ln:-Ki1BSib4Z4fLlKgcTM5:b-wlvkra) ([raw file](https://github.com/cockroachdb/cockroach/blob/10a642d72aacb5ac2a99f256977783d3bb510c5c/build/common.mk#L299)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Did you mean to remove this line before committing?
</blockquote></details>

I think not? It's still nice to see what it's doing.

---

*[build/push-tagged-aws.sh, line 53 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/14840#-Ki19jYItOtZteE84KtF:-Ki1Bb1iGZpPepmFO-1R:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/10a642d72aacb5ac2a99f256977783d3bb510c5c/build/push-tagged-aws.sh#L53)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

`build-osx.sh`? I think both references to that script need to point to the Makefile (or just remove these parts of the comments)
</blockquote></details>

Done.

---

*[build/teamcity-release-upload.sh, line 43 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/14840#-Ki1AAQikY36waljnuFI:-Ki1BVryrobCo7OqKfbJ:b3wgzfp) ([raw file](https://github.com/cockroachdb/cockroach/blob/10a642d72aacb5ac2a99f256977783d3bb510c5c/build/teamcity-release-upload.sh#L43)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Don't change it in this PR, but this part of the build process is taking long enough that we should look at parallelizing it soon.
</blockquote></details>

It should be much faster after this PR. `xgo` doesn't cache across runs, but our builder image does.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14840)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
Review status: 0 of 46 files reviewed at latest revision, 5 unresolved discussions, some commit checks failed.

---

*[Makefile, line 45 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/14840#-Ki15vq-1QQl82cSmc9U:-Ki1D8F_Ko6JDScKloSc:b-3nzwlg) ([raw file](https://github.com/cockroachdb/cockroach/blob/10a642d72aacb5ac2a99f256977783d3bb510c5c/Makefile#L45)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

should we rename both it and release-musl to `release-linux-gnu` and `release-linux-musl`?
</blockquote></details>

I'd keep `release-linux` for the glibc case because that is the "normal" linux release build. We can reconsider when/if musl becomes more of a first-class citizen.

---

*[build/common.mk, line 299 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/14840#-Ki17dnXa-uUeVRu38Ln:-Ki1CafsaB3Z5VPkgA_-:bo85d5l) ([raw file](https://github.com/cockroachdb/cockroach/blob/10a642d72aacb5ac2a99f256977783d3bb510c5c/build/common.mk#L299)):*
<details><summary><i>Previously, tamird (Tamir Duberstein) wrote…</i></summary><blockquote>

I think not? It's still nice to see what it's doing.
</blockquote></details>

Ah, I guess all the other rules print what they're doing but this one has `@` in front of everything (and removing that would be kind of silly). This is fine, although it might look a little nicer with `s/GEN/Generating/`. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14840)*
<!-- Sent from Reviewable.io -->

tamird commented:
Review status: 0 of 47 files reviewed at latest revision, 5 unresolved discussions.

---

*[Makefile, line 45 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/14840#-Ki15vq-1QQl82cSmc9U:-Ki1KiW3T53Kh3c8nCZh:bycuxi) ([raw file](https://github.com/cockroachdb/cockroach/blob/10a642d72aacb5ac2a99f256977783d3bb510c5c/Makefile#L45)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

I'd keep `release-linux` for the glibc case because that is the "normal" linux release build. We can reconsider when/if musl becomes more of a first-class citizen.
</blockquote></details>

I've gone with release-linux-gnu for explicitness. This doesn't leak out (the version command will still report `release`, for now).

---

*[build/common.mk, line 299 at r6](https://reviewable.io:443/reviews/cockroachdb/cockroach/14840#-Ki17dnXa-uUeVRu38Ln:-Ki1JbAikMuZDQ2aaG6N:b-j4gtrd) ([raw file](https://github.com/cockroachdb/cockroach/blob/10a642d72aacb5ac2a99f256977783d3bb510c5c/build/common.mk#L299)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

Ah, I guess all the other rules print what they're doing but this one has `@` in front of everything (and removing that would be kind of silly). This is fine, although it might look a little nicer with `s/GEN/Generating/`. 
</blockquote></details>

I think cmake prints `GEN`, which was the inspiration for this.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/14840)*
<!-- Sent from Reviewable.io -->

mjibson commented:
My workflow with `go test` works with this.

benesch committed:
build: make more consistent use of ORG_ROOT

BramGruneir commented:
Is there some other step besides checking out this branch and running make?  Do I need to install cmake?

I get
```
16:58 $ make
GOPATH set to /Users/bram/go
/usr/local/bin/bash: /Users/bram/go/src/github.com/cockroachdb/cockroach/bin/ncpus: No such file or directory
make[1]: Nothing to be done for `build_lib_static'.
mkdir -p /Users/bram/go/native/x86_64-apple-darwin16.5.0/protobuf
cd /Users/bram/go/native/x86_64-apple-darwin16.5.0/protobuf && cmake  -Dprotobuf_BUILD_TESTS=OFF /Users/bram/go/src/github.com/cockroachdb/cockroach/c-deps/protobuf.src/cmake
/usr/local/bin/bash: cmake: command not found
make: *** [/Users/bram/go/native/x86_64-apple-darwin16.5.0/protobuf/Makefile] Error 127
```

tamird commented:
Yes, you need cmake.

On Tue, Apr 18, 2017 at 4:59 PM, Bram Gruneir <notifications@github.com>
wrote:

> Is there some other step besides checking out this branch and running
> make? Do I need to install cmake?
>
> I get
>
> 16:58 $ make
> GOPATH set to /Users/bram/go
> /usr/local/bin/bash: /Users/bram/go/src/github.com/cockroachdb/cockroach/bin/ncpus: No such file or directory
> make[1]: Nothing to be done for `build_lib_static'.
> mkdir -p /Users/bram/go/native/x86_64-apple-darwin16.5.0/protobuf
> cd /Users/bram/go/native/x86_64-apple-darwin16.5.0/protobuf && cmake  -Dprotobuf_BUILD_TESTS=OFF /Users/bram/go/src/github.com/cockroachdb/cockroach/c-deps/protobuf.src/cmake
> /usr/local/bin/bash <http://github.com/cockroachdb/cockroach/c-deps/protobuf.src/cmake/usr/local/bin/bash>: cmake: command not found
> make: *** [/Users/bram/go/native/x86_64-apple-darwin16.5.0/protobuf/Makefile] Error 127
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/14840#issuecomment-294981050>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABdsPFNPejNUWUce_jK5OpgOBWHxk5oNks5rxSRAgaJpZM4M7L6s>
> .
>

BramGruneir commented:
All good, `go test` works perfectly.

dt commented:
Hm, just discovered this that when we untar c-deps/protobuf, it includes an "examples" folder which unfortunately includes some go files, and those go files import a non-existant package `:
```
$ go list -f '{{join .Imports "\n"}}' ./c-deps/...
...
github.com/google/protobuf/examples/tutorial
...
```

Since `github.com/google/protobuf/examples/tutorial` doesn't exist, `glide up` will fail when it tries to find and satisfy all imports in ./....

Should we add a protnbuf patch to nuke examples/? Or just nuke the folder in the makefile or the import script?
